### DISCUSSION TITLE
Can reMarkable use env proxy configurations?

### DISCUSSION THREAD

**Discussion by khose-ie** [toltec-dev/toltec#526] [Support] [answered]
Hi 

I have a question about reMarkable proxy.
Can reMarkable apps use internel environment proxy configurations like HTTP_PROXY to connect cloud ?

Because I am in China and can not connect reMarkable cloud, I want to install a shadowsocks-local by toltec in my tablet and set env HTTP_PROXY to make xochitl connect cloude via local proxy, will it work fine?

**Comment by LinusCDE** [comment, marked as answer]
I would think that this depends on the individual apps. Not sure what apps support it, but some might support it just by their used library.

For the rM Cloud, you could probably try to use [rmfakecloud](https://github.com/ddvk/rmfakecloud)'s approach to reroute the cloud traffic to a local http proxy and in use this to tunnel the requests through an additional proxy.

Otherwise the most straight forward way would probably be to set up an vpn. We have support for running wireguard on the device (installabale as a package). Innernet might make the wireguard usage easier for you if you find the traditional setup to in-depth. In case of any restrictions, you should usually also be able to download all the packages [from here](https://toltec-dev.org/stable/) and install the `.ipkg` files manually.

**Reply by khose-ie** [reply]
Thank you for you answer!

I  found that rmfakecloud's approach used the /etc/hosts file to redirect web url to local proxy, but I‘m afraid it will strip away the destination URL information.

I still need the destination URL information in my proxy tunnel to access the target web. 
So is there some good way to do proxy with URL information?

Thanks!

**Reply by LinusCDE** [reply]
It should work, if you can force the proxy to use a custom dns server / not use the standard library that respects `/etc/hosts`. This should be the only information that is missing. The original domain is passed in a header that the proxy should be able to use.

**Reply by khose-ie** [reply]
Thank you! But you mean that I need to install a custom DNS server in reMarkable?

**Reply by LinusCDE** [reply]
Not exactly. The thing is that every OS also caches typicial DNS requests made (in linux this might even be a part of the namingstuff, not really sure here). This is also used to inject custom entries using the host file. If xochitl has the request redirected to a local web server, but the webserver gets its own ip for requests, it would not work at all (= endless loop). So the local proxy server needs to specifically request some DNS server to bypass linux overriding the entry with the one from the host file.

**Reply by khose-ie** [reply]
I try to use environment variables and it worked fine! Thank your for your guidance.

**Comment by khose-ie** [comment]
@LinusCDE Hi, unfortunately, now my cloud is work, but the button "Send by email" is disable, it seems this program is not used the proxy, it may be not in xochitl ? So do you konw which service control the "Send by email" function?

**Reply by LinusCDE** [reply]
Not sure myself. Maybe the rmfakecloud repo has some documentation on what the difference is (you can set it up your self afaik). @ddvk do you know what could cause the emails to not work?

**Reply by ddvk** [reply]
i'm not sure, emails use the same api. unless the websocket is not connected and it shows no cloud or something with the subscription model

**Reply by khose-ie** [reply]
@ddvk Ok, all right, thank your and I go to use your awesome project rmfakecloud now.
